java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/SearchTreeBasedDictionaryBuilder.java

```diff
@@ -0,0 +1,202 @@
+ * the dictionary is not known a priori, so it is generated dynamically.
+ * In particular, when a new value arrives, it is tested to check if it is already
+ * in the dictionary. If so, it is simply neglected, otherwise, it is added to the dictionary.
+ *
+ * <p>
+ *   This class builds the dictionary based on a binary search tree.
+ *   Each add operation can be finished in O(log(n)) time,
+ *   where n is the current dictionary size.
+ * </p>
+ * <p>
+ *   The dictionary builder is intended to build a single dictionary.
+ *   So it cannot be used for different dictionaries.
+ * </p>
+ * <p>Below gives the sample code for using the dictionary builder
+ * <pre>{@code
+ * SearchTreeBasedDictionaryBuilder dictionaryBuilder = ...
+ * dictionaryBuilder.startBuild();
+ * ...
+ * dictionaryBuild.addValue(newValue);
+ * ...
+ * dictionaryBuilder.endBuild();
+ * }</pre>
+ * </p>
+ * <p>
+ *  With the above code, the dictionary vector will be populated,
+ *  and it can be retrieved by the {@link SearchTreeBasedDictionaryBuilder#getDictionary()} method.
+ *  After that, dictionary encoding can proceed with the populated dictionary encoder.
+ * </p>
+ * @param <V> the dictionary vector type.
+ */
+public class SearchTreeBasedDictionaryBuilder<V extends ValueVector> {
+
+  /**
+   * The dictionary to be built.
+   */
+  private final V dictionary;
+
+  /**
+   * The criteria for sorting in the search tree.
+   */
+  protected final VectorValueComparator<V> comparator;
+
+  /**
+   * The total number of data items probed.
+   */
+  private int totalValueCount;
+
+  /**
+   * The number of items in the dictionary.
+   */
+  private int distinctValueCount;
+
+  /**
+   * If null should be encoded.
+   */
+  private final boolean encodeNull;
+
+  /**
+   * The search tree for storing the value index.
+   */
+  private TreeSet<Integer> searchTree;
+
+  /**
+   * A flag indicating if dictionary building is finished.
+   */
+  private boolean buildFinished = false;
+
+  /**
+   * Construct a search tree-based dictionary builder.
+   * @param dictionary the dictionary vector.
+   * @param comparator the criteria for value equality.
+   */
+  public SearchTreeBasedDictionaryBuilder(V dictionary, VectorValueComparator<V> comparator) {
+    this(dictionary, comparator, false);
+  }
+
+  /**
+   * Construct a search tree-based dictionary builder.
+   * @param dictionary the dictionary vector.
+   * @param comparator the criteria for value equality.
+   * @param encodeNull if null should be encoded.
+   */
+  public SearchTreeBasedDictionaryBuilder(V dictionary, VectorValueComparator<V> comparator, boolean encodeNull) {
+    this.dictionary = dictionary;
+    this.comparator = comparator;
+    this.encodeNull = encodeNull;
+    this.comparator.attachVector(dictionary);
+
+    searchTree = new TreeSet<>((index1, index2) -> comparator.compare(index1, index2));
+  }
+
+  /**
+   * Start building the dictionary.
+   */
+  public void startBuild() {
```

Comment:
is it important to keep the state of startBuild and endBuild. Is seems like we can set ValueCount whenever getDictionary is called if users want an incremental dictionary?

Reply:
Sounds good. Revised accordingly. Thanks.